substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -94,17 +94,246 @@ private PosixJavaNetSubstitutions() {
 @TargetClass(className = "java.net.PlainDatagramSocketImpl")
 @Platforms({Platform.LINUX.class, Platform.DARWIN.class})
 final class Target_java_net_PlainDatagramSocketImpl {
+    @Alias boolean connected;
 
     @Substitute
     @SuppressWarnings({"static-method", "unused"})
     protected void bind0(int lport, InetAddress laddr) throws SocketException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.PlainDatagramSocketImpl.bind0(int, InetAddress)");
+        //        jobject fdObj = (*env)->GetObjectField(env, this, pdsi_fdID);
+        //        int fd;
+        //        int len = 0;
+        CIntPointer len_Pointer = StackValue.get(CIntPointer.class);
+        //        SOCKADDR him;
+        Socket.sockaddr him = StackValue.get(JavaNetNetUtilMD.SOCKADDR_LEN());
+        //
+        FileDescriptor fdObj = Util_java_net_DatagramSocketImpl.as_Target_java_net_DatagramSocketImpl(this).fd;
+        //        if (IS_NULL(fdObj)) {
+        if (fdObj == null) {
+            throw new SocketException("Socket closed");
+        }
+        //            JNU_ThrowByName(env, JNU_JAVANETPKG "SocketException",
+        //                            "Socket closed");
+        //            return;
+        //        } else {
+        int fd = PosixUtils.getFD(fdObj);
```

Comment:
Tempting as it is to simplify the control flow (changing `if () { throw } else ....` to `if () { throw }; ....` that is one more thing I have to think about when reviewing the code.

To be fair, I usually reuse `} else {` and `}` lines from the original C code.  But I try not to change the control flow unless I have to.  Among the reasons for changing the control flow: In C code one can "throw" and exception and then continue execution, knowing that the exception will not be noticed until one crosses back from C to Java.  In SubstrateVM it is all Java code, so if you throw and exception, execute terminate abruptly right there.  That's a rare case.  Sometimes I rewrite the code, sometimes I wrap `try ... finally ...` blocks around things.